client/write.go

```diff
@@ -131,31 +136,29 @@ func (w *writeState) close() {
 }
 
 func (w *writeState) completionFn(result interface{}, err error) {
-	remaining := atomic.AddInt32(&w.pending, -1)
-	successful := int32(0)
-	if err == nil {
-		successful = atomic.AddInt32(&w.successful, 1)
-	}
+	hostID := result.(topology.Host).ID()
+	// NB(bl) panic on invalid result, it indicates a bug in the code
 
 	w.Lock()
+	w.pending--
 
 	if err != nil {
 		w.errors = append(w.errors, err)
+	} else if hostShardSet, ok := w.topoMap.LookupHostShardSet(hostID); !ok {
+		w.errors = append(w.errors, fmt.Errorf("Missing host shard in writeState completionFn: %s", hostID))
+	} else if shardState, err :=
+		hostShardSet.ShardSet().LookupStateByID(w.op.shardID); err != nil {
+		w.errors = append(w.errors, fmt.Errorf("Missing shard %d in host %s", w.op.shardID, hostID))
+	} else if shardState == shard.Available {
+		// NB(bl): only count writes to available shards towards success
+		w.success++
 	}
 
-	switch w.session.writeLevel {
-	case topology.ConsistencyLevelOne:
-		if successful > 0 || remaining == 0 {
-			w.Signal()
-		}
-	case topology.ConsistencyLevelMajority:
-		if successful >= w.majority || remaining == 0 {
-			w.Signal()
-		}
-	case topology.ConsistencyLevelAll:
-		if remaining == 0 {
-			w.Signal()
-		}
+	wLevel := w.session.writeLevel
+	if w.pending == 0 ||
```

Comment:
I don't think there's a delay, the two versions should be logically equivalent. Did I mess something up?